toolchain/check/decl_state.h

```diff
@@ -0,0 +1,80 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_CHECK_DECL_STATE_H_
+#define CARBON_TOOLCHAIN_CHECK_DECL_STATE_H_
+
+namespace Carbon::Check {
+
+class KeywordModifierSet {
+ public:
+  enum RawEnum {
+    // At most one of these access modifiers allowed for a given declaration,
+    // and if present it must be first:
+    Private = 1 << 0,
+    Protected = 1 << 1,
+
+    // At most one of these declaration modifiers allowed for a given
+    // declaration:
+    Abstract = 1 << 2,
+    Base = 1 << 3,
+    Default = 1 << 4,
+    Final = 1 << 5,
+    Impl = 1 << 6,
+    Virtual = 1 << 7,
+
+    // Sets of modifiers:
+    Access = Private | Protected,
+  };
+
+  KeywordModifierSet() : keywords(static_cast<RawEnum>(0)) {}
+  KeywordModifierSet(RawEnum k) : keywords(k) {}
+  auto Set(RawEnum to_set) const -> KeywordModifierSet {
+    KeywordModifierSet ret;
+    ret.keywords = static_cast<RawEnum>(keywords | to_set);
+    return ret;
+  }
+  auto SetPrivate() const -> KeywordModifierSet { return Set(Private); }
+  auto SetProtected() const -> KeywordModifierSet { return Set(Protected); }
+  auto SetAbstract() const -> KeywordModifierSet { return Set(Abstract); }
+  auto SetBase() const -> KeywordModifierSet { return Set(Base); }
+  auto SetDefault() const -> KeywordModifierSet { return Set(Default); }
+  auto SetFinal() const -> KeywordModifierSet { return Set(Final); }
+  auto SetImpl() const -> KeywordModifierSet { return Set(Impl); }
+  auto SetVirtual() const -> KeywordModifierSet { return Set(Virtual); }
+
+  auto Has(unsigned to_Has) const -> bool { return keywords & to_Has; }
```

Comment:
```suggestion
  auto Has(unsigned has_mask) const -> bool { return keywords & has_mask; }
```

"Has" -> "has" is for style.

"to_has" -> "has_mask" is a naming suggestion: I'm not sure "to" means for "has"; "has_mask" or similar might emphasize that this is a mask check.

Reply:
Code has been replaced.